### PR TITLE
implementation of precomputed sqrt algo

### LINKED ISSUE #49: use Tonelli-Shanks with precomputed tables for proof deserialization
Description: https://ihagopian.com/posts/tonelli-shanks-with-precomputed-dlog-tables

Impl1: https://github.com/GottfriedHerold/Bandersnatch/blob/f665f90b64892b9c4c89cff3219e70456bb431e5/bandersnatch/fieldElements/field_element_square_root.go
Impl2:https://github.com/crate-crypto/go-ipa/pull/38

### PR DESCRIPTION
**Pull request by dkgoutham**
Closes #49

### CODE REVIEW

File: src/Nethermind.Verkle/Fields/FpEElement/Element.Sqrt.cs [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        x0 = mapG2_24[fourthEq];
+
+        ulong xBy2 = ((ulong)1<<23)*x0 + ((ulong)1<<15)*x1 + ((ulong)1<<7)*x2 + x3/2;
+        xBy2 = ((ulong)1<<32) - xBy2;
+        computePower(xBy2, table, out final);
+        MultiplyMod(final, squareRootCandidate, out final);
+    }
+
+    public static void TestSqrt(){
+        using IEnumerator<FE> set = GetRandom().GetEnumerator();
```
Comment (tanishqjasoria) [inline src/Nethermind.Verkle/Fields/FpEElement/Element.Sqrt.cs:279]: move test to corresponding Verkle.Tests

File: src/Nethermind.Verkle/Fields/FpEElement/Element.Sqrt.cs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    //inverse of dyadicRootOfUnity, that is inverse of 2^32 th primitive root.
+    public static readonly FE gInv = new(gInv_0, gInv_1, gInv_2, gInv_3);
+
+    public static FE[,] GenerateLookUpTable(){
+        FE[,] table = new FE[4, 256];
+        FE[] gValues = {dyadicRootofUnity, g8, g16, g24};
+
+        for(int i=0;i<4;i++){
+            table[i,0] = FE.One;
+            for(int j=1;j<256;j++){
```
Comment (tanishqjasoria) [inline src/Nethermind.Verkle/Fields/FpEElement/Element.Sqrt.cs:57]: parallelize the outer loop, the inner loop only depends on i

File: src/Nethermind.Verkle/Fields/FpEElement/Element.Sqrt.cs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+            }else if(i==15){
+                arr[2] = res;
+            }
+        }
+        arr[3] = res;
+        return arr;
+    }
+
+//Given a number x, this method returns y0, y1, y2, y3: where x = (2^24)*y0 + (2^16)*y1 + (2^8)*y2 + (2^0)*y3
+    public static ulong[] decomposeNumber(ulong x){
```
Comment (tanishqjasoria) [inline src/Nethermind.Verkle/Fields/FpEElement/Element.Sqrt.cs:219]: use byte[] to represent the decomposed number

**Review by tanishqjasoria — CHANGES_REQUESTED** [review]
also add and compare these two implementations in Verkle.Bench
comparing using Stopwatch is inaccurate

File: src/Nethermind.Verkle/Fields/FpEElement/Element.Sqrt.cs (hunk 3 above)
Comment (dkgoutham) [reply src/Nethermind.Verkle/Fields/FpEElement/Element.Sqrt.cs:219]: public static byte[] decomposeNumber(ulong x){
    byte y0 = (byte)((x >> 24) & 0xFF); // Shift right by 24 bits and mask out the last 8 bits, then cast to byte
    byte y1 = (byte)((x >> 16) & 0xFF); // Shift right by 16 bits and mask out the last 8 bits, then cast to byte
    byte y2 = (byte)((x >> 8) & 0xFF);  // Shift right by 8 bits and mask out the last 8 bits, then cast to byte
    byte y3 = (byte)(x & 0xFF);         // Mask out the last 8 bits and cast to byte
    byte[] arr = {y0, y1, y2, y3}; 
    return arr;
}

Is this what you mean?

**Pull request closed without merging**